main.cpp

```diff
@@ -271,322 +224,180 @@ struct HeapA
     A* aPtr;
 };
 
-#include <iostream>
-#include <cmath>
 
-struct DoubleType;
-struct IntType;
 
-struct FloatType
+template<typename NumericType>
+struct Numeric
 {
-    FloatType(float floatIn_) : value(new float(floatIn_)){}
+    using Type = NumericType;
     
-    ~FloatType()
+    Numeric(Type valuePtrIn_) : valuePtr(std::make_unique<Type>(valuePtrIn_)){}
+     
+    ~Numeric()
     {
-        delete value;
-        value = nullptr;
+        valuePtr = nullptr;
     }
+
+    operator Type() const { return *valuePtr; }
     
-    operator float() const { return *value;}
-    
-    FloatType& operator +=(float rhs); 
-    FloatType& operator -=(float rhs);  
-    FloatType& operator *=(float rhs);
-    FloatType& operator /=(float rhs);
-  
-    float pow(const IntType&);
-    float pow(const FloatType&);
-    float pow(const DoubleType&);
-    float pow(float);
+    Numeric& operator +=(const Type rhs) 
+    {
+        *valuePtr += rhs; 
+        return *this;
+    }
+    Numeric& operator -=(const Type rhs) 
+    {
+        *valuePtr -= rhs; 
+        return *this;
+    }
+    Numeric& operator *=(const Type rhs) 
+    {
+        *valuePtr *= rhs; 
+        return *this;
+    }
+    Numeric& operator /=(const Type rhs)
+    {
+        if constexpr (std::is_same<Type, int>::value)
+        {
+            if constexpr (std::is_same<decltype(rhs), const int>::value)
+            {
+                if (rhs == 0)
+                {
+                    std::cout << "can't divide integers by zero!" <<std::endl;
+                    return *this;
+                }
+            }
+            else if (rhs < std::numeric_limits<Type>::epsilon())
+            {
+                std::cout << "can't divide integers by zero!" <<std::endl;
+                return *this;   
+            }    
+        }
+        else if (rhs < std::numeric_limits<Type>::epsilon())
+        {
+            std::cout << "warning: floating point division by zero!" <<std::endl;           
+        }
+        *valuePtr /= rhs;
+        return *this;   
+    }
     
-    float powInternal(float* val, float arg);
+    Numeric& pow(Type exponent)
+    {
+        return powInternal( exponent);
+    }
+   
+    Numeric& powInternal(Type arg)
+    {
+        *valuePtr = static_cast<Type>(std::pow( *valuePtr, arg));
+        return *this;
+    }
 
-    FloatType& apply(std::function<FloatType&(float&)> func)
+    Numeric& apply(std::function<Numeric&( std::unique_ptr<Type>& )> func)
     { 
-        return func ? func(*value) : *this;      
+        if(func)
+             return func(valuePtr);
+        return *this;
     }
 
-    FloatType& apply(void(*func)(float&))
+    Numeric& apply(void(*func)( std::unique_ptr<Type>&) )
     {
         if(func)
-            func(*value);
+            func(valuePtr);
         return *this;
     }
      
 private:
-    float* value;
+    std::unique_ptr<Type> valuePtr;
```

Comment:
change this name back to `value`.
There was no need to rename the variable.   all you did by doing that was create more work for yourself.
- [ ] revise

Reply:
I changed it when I was debugging as was worried about conflicts with the static variable 'value' in is_same. Seemed pointless at the time tbh.